### DISCUSSION TITLE
WiFi.localIP() hangs in Nano RP2040 Connect with Arduino-Pico core (EarlePhilhower)

### DISCUSSION THREAD

**Discussion by stefat** [khoih-prog/WiFiNINA_Generic#24] [Q&A]
**Settings:**
- Hw Mac Apple Silicon M1Pro 
- macOS Ventura 13.0 
- Sketch: Examples->WiFiNINA_Generic->WiFiWebServer
- Arduino IDE 1.8.19
Nano RP2040 Connect sdk core: Arduino-Pico core in ~ /Library/Arduino15/packages/rp2040/2.6.2/
Libraries: 
- WiFiNINA_Generic 1.8.14-6 
- SPI 1.0 - from ~/Library/Arduino15/packages/rp2040/hardware/rp2040/2.6.2/libraries/SPI

**The Issue:**
Just successfully connected the sw hangs doing a IPAddress ip = WiFi.localIP() instruction.
The RP2040 connected as a Station or AP shows the same behaviour. 

**The code:**

The following printWiFiStatus() is called: edited from the original example sketch.

```cpp
void printWiFiStatus() 
{
  // print the SSID of the network you're attached to:
  Serial.print(F("SSID: "));
  Serial.println(WiFi.SSID());
  delay(1000);
  // print your board's IP address:
  Serial.print(F("IP load "));
  IPAddress ip = WiFi.localIP();
  Serial.print(F("IP Address returned:"));
  Serial.println(ip);

  // print the received signal strength:
  long rssi = WiFi.RSSI();
  Serial.print(F("Signal strength (RSSI):"));
  Serial.print(rssi);
  Serial.println(F(" dBm"));
}`
```

The serial monitor shows:

```
18:01:17.764 -> Start WiFiWebServer on NANO_RP2040_CONNECT
18:01:17.764 -> WiFiNINA_Generic v1.8.14-6
18:01:18.488 -> Attempting to connect to SSID: TIM-86351700
18:01:38.388 -> SSID: TIM-86351700
18:01:39.373 -> IP load 
```

and nothing more. 

Any clue ?
Thanks for any help.

**Comment by khoih-prog** [comment]
HI @stefat 

Thanks for your interest in the library.

I've tested and everything is OK using the [**WiFiWebClient** example](https://github.com/khoih-prog/WiFiNINA_Generic/tree/master/examples/WiFiWebClient), which include the similar code

https://github.com/khoih-prog/WiFiNINA_Generic/blob/db7af0057bbc0318c1ec33f710b44bce5ad3209e/examples/WiFiWebClient/WiFiWebClient.ino#L149-L165

Try to use the latest `arduino-pico` core v2.6.3 or Arduino `mbed_nano` core v3.4.1

---

#### Debug Terminal

```
Start WiFiWebClient on NANO_RP2040_CONNECT
WiFiNINA_Generic v1.8.14-6
Attempting to connect to SSID: HueNet1
Connected to WiFi
SSID: HueNet1
IP Address: 192.168.2.117
Signal strength (RSSI):-21 dBm

Starting connection to server...
Connected to server
HTTP/1.1 200 OK
Date: Fri, 11 Nov 2022 18:04:19 GMT
Content-Type: text/plain
Content-Length: 2263
Connection: close
x-amz-id-2: WlCZ31rmgvCEyPn/LO8PATIOj9GGaqtJQs7RqLxrzyyPmjeyNA+z00xL30lyiTYf5pgSnpQyYoc=
x-amz-request-id: 310ZJVPNXA3SKR92
Last-Modified: Wed, 23 Feb 2022 14:56:42 GMT
ETag: "667cf48afcc12c38c8c1637947a04224"
CF-Cache-Status: DYNAMIC
Report-To: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=A5gcrNb58k3esdyICxZ6ikTGqg2%2FmoNfT0s%2FbbHRYM8LpEIoJIfCFoQ0qQAO5smx8NWASH%2FL3LJ4Hk%2FZcYOrdDSTjamXSz9H9avlcihUJ%2BmGglZ0mLd4PPYwo0XdUEA%3D"}],"group":"cf-nel","max_age":604800}
NEL: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
Server: cloudflare
CF-RAY: 7688f31bdf34179d-EWR
alt-svc: h3=":443"; ma=86400, h3-29=":443"; ma=86400


           `:;;;,`                      .:;;:.           
        .;;;;;;;;;;;`                :;;;;;;;;;;:     TM 
      `;;;;;;;;;;;;;;;`            :;;;;;;;;;;;;;;;      
     :;;;;;;;;;;;;;;;;;;         `;;;;;;;;;;;;;;;;;;     
    ;;;;;;;;;;;;;;;;;;;;;       .;;;;;;;;;;;;;;;;;;;;    
   ;;;;;;;;:`   `;;;;;;;;;     ,;;;;;;;;.`   .;;;;;;;;   
  .;;;;;;,         :;;;;;;;   .;;;;;;;          ;;;;;;;  
  ;;;;;;             ;;;;;;;  ;;;;;;,            ;;;;;;. 
 ,;;;;;               ;;;;;;.;;;;;;`              ;;;;;; 
 ;;;;;.                ;;;;;;;;;;;`      ```       ;;;;;`
 ;;;;;                  ;;;;;;;;;,       ;;;       .;;;;;
`;;;;:                  `;;;;;;;;        ;;;        ;;;;;
,;;;;`    `,,,,,,,,      ;;;;;;;      .,,;;;,,,     ;;;;;
:;;;;`    .;;;;;;;;       ;;;;;,      :;;;;;;;;     ;;;;;
:;;;;`    .;;;;;;;;      `;;;;;;      :;;;;;;;;     ;;;;;
.;;;;.                   ;;;;;;;.        ;;;        ;;;;;
 ;;;;;                  ;;;;;;;;;        ;;;        ;;;;;
 ;;;;;                 .;;;;;;;;;;       ;;;       ;;;;;,
 ;;;;;;               `;;;;;;;;;;;;                ;;;;; 
 `;;;;;,             .;;;;;; ;;;;;;;              ;;;;;; 
  ;;;;;;:           :;;;;;;.  ;;;;;;;            ;;;;;;  
   ;;;;;;;`       .;;;;;;;,    ;;;;;;;;        ;;;;;;;:  
    ;;;;;;;;;:,:;;;;;;;;;:      ;;;;;;;;;;:,;;;;;;;;;;   
    `;;;;;;;;;;;;;;;;;;;.        ;;;;;;;;;;;;;;;;;;;;    
      ;;;;;;;;;;;;;;;;;           :;;;;;;;;;;;;;;;;:     
       ,;;;;;;;;;;;;;,              ;;;;;;;;;;;;;;       
         .;;;;;;;;;`                  ,;;;;;;;;:         
                                                         
                                                         
                                                         
                                                         
    ;;;   ;;;;;`  ;;;;:  .;;  ;; ,;;;;;, ;;. `;,  ;;;;   
    ;;;   ;;:;;;  ;;;;;; .;;  ;; ,;;;;;: ;;; `;, ;;;:;;  
   ,;:;   ;;  ;;  ;;  ;; .;;  ;;   ,;,   ;;;,`;, ;;  ;;  
   ;; ;:  ;;  ;;  ;;  ;; .;;  ;;   ,;,   ;;;;`;, ;;  ;;. 
   ;: ;;  ;;;;;:  ;;  ;; .;;  ;;   ,;,   ;;`;;;, ;;  ;;` 
  ,;;;;;  ;;`;;   ;;  ;; .;;  ;;   ,;,   ;; ;;;, ;;  ;;  
  ;;  ,;, ;; .;;  ;;;;;:  ;;;;;: ,;;;;;: ;;  ;;, ;;;;;;  
  ;;   ;; ;;  ;;` ;;;;.   `;;;:  ,;;;;;, ;;  ;;,  ;;;;   

Disconnecting from server.
```

**Comment by khoih-prog** [comment]
Hi @stefat 

I just found out that if you run [**WiFiWebServer** example](https://github.com/khoih-prog/WiFiNINA_Generic/tree/master/examples/WiFiWebServer) with `arduino-pico` core, these `analogRead()` calls will hang the system. This is still working in Arduino `mbed_nano` core

https://github.com/khoih-prog/WiFiNINA_Generic/blob/db7af0057bbc0318c1ec33f710b44bce5ad3209e/examples/WiFiWebServer/WiFiWebServer.ino#L134-L143

You have to remove them for that  `arduino-pico` core.

You can help open an issue in the `arduino-pico` core about this strange `anaogRead()` bug

**Comment by khoih-prog** [comment]
Hi @stefat 

Check the new release [**WiFiNINA_Generic** v1.8.14-7](https://github.com/khoih-prog/WiFiNINA_Generic/releases/tag/v1.8.14-7). Your contribution is noted in [**Contributions and Thanks**](https://github.com/khoih-prog/WiFiNINA_Generic#contributions-and-thanks)

No issue or discussion is too small to be ignored. Please feel free to post any problem next time, certainly after reasonable tests and retests. The library will be better and better thanks to **contributing users** like you.

Best Regards,

---

### Releases v1.8.14-7

1. Modify [**WiFiWebServer**](https://github.com/khoih-prog/WiFiNINA_Generic/tree/master/examples/WiFiWebServer) example to avoid `analogRead()` crash in `arduino-pico` core. Check [WiFi.localIP() hangs in Nano RP2040 Connect with Arduino-Pico core (EarlePhilhower) #24](https://github.com/khoih-prog/WiFiNINA_Generic/discussions/24)
2. Use `allman astyle` and add `utils`
3. Update `Packages' Patches`

**Reply by stefat** [reply]
Hello @khoih-prog,
I tested again under same settings but the WiFinina_Generic ver. 1.8.14-7 instead of 6. Same result: localIP hangs.
See please below:
07:10:14.343 -> Start WiFiWebServer on NANO_RP2040_CONNECT
07:10:14.343 -> WiFiNINA_Generic v1.8.14-7
07:10:15.072 -> Attempting to connect to SSID: TIM-86351700
07:10:19.856 -> SSID: TIM-86351700
07:10:20.869 -> IP load

**Comment by khoih-prog** [comment]
Did you remove

https://github.com/khoih-prog/WiFiNINA_Generic/blob/db7af0057bbc0318c1ec33f710b44bce5ad3209e/examples/WiFiWebServer/WiFiWebServer.ino#L134-L143

and replace with

https://github.com/khoih-prog/WiFiNINA_Generic/blob/bbce55c3d94f76cbb063fe277ee9ac050db6438e/examples/WiFiWebServer/WiFiWebServer.ino#L141-L153

You can't use `analogRead()` for **Nano_RP2040_Connect** in arduino-pico core now.

**Reply by stefat** [reply]
not yet. I am testing under mbed core: it works.
07:29:03.448 -> Start WiFiWebServer on MBED NANO_RP2040_CONNECT
07:29:03.448 -> WiFiNINA_Generic v1.8.14-7
07:29:04.198 -> Attempting to connect to SSID: TIM-86351700
07:29:13.081 -> SSID: TIM-86351700
07:29:14.075 -> IP load IP Address returned:192.168.1.9
07:29:14.075 -> Signal strength (RSSI):-87 dBm

I try now removing analogRead() on both platforms

**Comment by stefat** [comment]
yes ! Problem is analogRead() with Arduino-Pico.
Your patch removing analogRead() works on both cores: mbed and Arduino-Pico.
BTW your patch misses an 'int'. I used this:

```
for (int analogChannel = 0; analogChannel < 6; analogChannel++)
{
  // analogRead() will hang system in arduino-pico core
   int sensorReading = millis() % 1024;      //analogRead(analogChannel); 
   client.print("analog input ");
   client.print(analogChannel);
   client.print(" is ");
   client.print(sensorReading);
   client.println("<br />");
}
```

Serial monitor on both platforms:

```
7:37:56.543 -> Start WiFiWebServer on MBED NANO_RP2040_CONNECT
07:37:56.543 -> WiFiNINA_Generic v1.8.14-7
07:37:57.297 -> Attempting to connect to SSID: TIM-86351700
07:38:01.875 -> SSID: TIM-86351700
07:38:02.865 -> IP load IP Address returned:192.168.1.9
07:38:02.865 -> Signal strength (RSSI):-86 dBm
07:39:11.855 -> 
```

```
07:39:11.855 -> Start WiFiWebServer on NANO_RP2040_CONNECT
07:39:11.855 -> WiFiNINA_Generic v1.8.14-7
07:39:12.617 -> Attempting to connect to SSID: TIM-86351700
07:39:14.913 -> Attempting to connect to SSID: TIM-86351700
07:39:16.995 -> Attempting to connect to SSID: TIM-86351700
07:39:25.550 -> SSID: TIM-86351700
07:39:26.535 -> IP load IP Address returned:192.168.1.9
07:39:26.535 -> Signal strength (RSSI):-85 dBm
```

**Reply by khoih-prog** [reply]
It's there in the example

https://github.com/khoih-prog/WiFiNINA_Generic/blob/bbce55c3d94f76cbb063fe277ee9ac050db6438e/examples/WiFiWebServer/WiFiWebServer.ino#L141-L153

**Reply by stefat** [reply]
I missed that :-( ... sorry

**Reply by khoih-prog** [reply]
You can help open an issue in the `arduino-pico` core about this strange `analogRead()` bug

**Reply by stefat** [reply]
I do
